### PR TITLE
Deprecate key labels of rb/crb

### PR DESCRIPTION
**Pull request by whitewindmills**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
```
resourcebinding.karmada.io/key -> resourcebinding.karmada.io/permanent-id
clusterresourcebinding.karmada.io/key -> clusterresourcebinding.karmada.io/permanent-id
```

**Which issue(s) this PR fixes**:
part of https://github.com/karmada-io/karmada/issues/4711

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
`karmada-controller-manager`: The labels `resourcebinding.karmada.io/key` and `clusterresourcebinding.karmada.io/key` has been deprecated and not be populated in Work objects(replaced by `resourcebinding.karmada.io/permanent-id` and `clusterresourcebinding.karmada.io/permanent-id` respectively).
```

### CODE REVIEW

**Comment by whitewindmills** [comment]
Split from https://github.com/karmada-io/karmada/pull/4743.
/cc @RainbowMango

**Comment by RainbowMango** [comment]
> Split from #4743. /cc @RainbowMango

Nice, I like small PRs. Thanks. I'll work on this then #4743.

**Comment by RainbowMango** [comment]
Re-run the test as [the failure seems not related](https://github.com/karmada-io/karmada/actions/runs/8598835846/job/23560475384?pr=4811#step:5:616):
> Error: operator/pkg/util/util.go:20:2: read /opt/hostedtoolcache/go/1.21.8/x64/src/archive/tar/example_test.go: unexpected NUL in input

File: pkg/apis/work/v1alpha1/well_known_constants.go [hunk 1]
```diff
@@ -16,19 +16,25 @@ limitations under the License.
...
 	// ResourceBindingNameLabel is added to objects to specify associated ResourceBinding's name.
+	// Deprecated
 	ResourceBindingNameLabel = "resourcebinding.karmada.io/name"
 
 	// ClusterResourceBindingLabel is added to objects to specify associated ClusterResourceBinding.
+	// Deprecated
 	ClusterResourceBindingLabel = "clusterresourcebinding.karmada.io/name"
 
 	// WorkNamespaceLabel is added to objects to specify associated Work's namespace.
+	// Deprecated
```
Comment (XiShanYongYe-Chang) [inline pkg/apis/work/v1alpha1/well_known_constants.go:34]: Can work-related labels be split into a separate pr?

Comment (whitewindmills) [reply pkg/apis/work/v1alpha1/well_known_constants.go:34]: In order to process PRs in parallel, this PR just marks work-related labels as deprecated, handling work-related labels will be in https://github.com/karmada-io/karmada/pull/4813.

File: pkg/util/helper/work.go [hunk 2]
```diff
@@ -114,37 +110,16 @@ func GetWorksByLabelsSet(c client.Client, ls labels.Set) (*workv1alpha1.WorkList
...
-				util.GetAnnotationValue(workList.Items[i].GetAnnotations(), workv1alpha2.ResourceBindingNamespaceAnnotationKey) == bindingNamespace {
-				retWorkList.Items = append(retWorkList.Items, workList.Items[i])
-			}
-		} else { // filter Works that derived by 'ClusterResourceBinding'
-			if util.GetAnnotationValue(workList.Items[i].GetAnnotations(), workv1alpha2.ClusterResourceBindingAnnotationKey) == bindingName {
-				retWorkList.Items = append(retWorkList.Items, workList.Items[i])
-			}
-		}
+		ls = labels.Set{workv1alpha2.ClusterResourceBindingPermanentIDLabel: bindingID}
 	}
```
Comment (RainbowMango) [inline pkg/util/helper/work.go:120]: Glad to see this removed!

File: pkg/dependenciesdistributor/dependencies_distributor.go [hunk 3]
```diff
@@ -522,10 +522,6 @@ func (d *DependenciesDistributor) createOrUpdateAttachedBinding(attachedBinding
 	key := client.ObjectKeyFromObject(attachedBinding)
 	err := d.Client.Get(context.TODO(), key, existBinding)
 	if err == nil {
-		if util.GetLabelValue(existBinding.Labels, workv1alpha2.ResourceBindingPermanentIDLabel) == "" {
-			existBinding.Labels = util.DedupeAndMergeLabels(existBinding.Labels,
-				map[string]string{workv1alpha2.ResourceBindingPermanentIDLabel: uuid.New().String()})
-		}
```
Comment (RainbowMango) [inline pkg/dependenciesdistributor/dependencies_distributor.go:528]: Why remove this? Can we ensure that the `existBinding` has the label?

File: pkg/dependenciesdistributor/dependencies_distributor.go [hunk 4]
```diff
@@ -544,7 +540,7 @@ func (d *DependenciesDistributor) createOrUpdateAttachedBinding(attachedBinding
 
 	attachedBinding.Labels = util.DedupeAndMergeLabels(attachedBinding.Labels,
 		map[string]string{workv1alpha2.ResourceBindingPermanentIDLabel: uuid.New().String()})
-	if err := d.Client.Create(context.TODO(), attachedBinding); err != nil {
+	if err := d.Client.Create(context.TODO(), attachedBinding); err != nil && !apierrors.IsAlreadyExists(err) {
```
Comment (RainbowMango) [inline pkg/dependenciesdistributor/dependencies_distributor.go:543]: Can you remind me when will rise the `already exist ` error?

File: pkg/controllers/binding/common.go [hunk 5]
```diff
@@ -157,18 +157,20 @@ func mergeLabel(workload *unstructured.Unstructured, binding metav1.Object, scop
...
 	if scope == apiextensionsv1.NamespaceScoped {
 		bindingID := util.GetLabelValue(binding.GetLabels(), workv1alpha2.ResourceBindingPermanentIDLabel)
-		util.MergeLabel(workload, workv1alpha2.ResourceBindingReferenceKey, names.GenerateBindingReferenceKey(binding.GetNamespace(), binding.GetName()))
 		util.MergeLabel(workload, workv1alpha2.ResourceBindingPermanentIDLabel, bindingID)
+		workLabel[workv1alpha2.ResourceBindingPermanentIDLabel] = bindingID
 
+		// TODO(whitewindmills): Delete the following two lines in a future version.
 		workLabel[workv1alpha2.ResourceBindingReferenceKey] = names.GenerateBindingReferenceKey(binding.GetNamespace(), binding.GetName())
-		workLabel[workv1alpha2.ResourceBindingPermanentIDLabel] = bindingID
+		util.MergeLabel(workload, workv1alpha2.ResourceBindingReferenceKey, names.GenerateBindingReferenceKey(binding.GetNamespace(), binding.GetName()))
```
Comment (RainbowMango) [inline pkg/controllers/binding/common.go:165]: a future version, do you mean v1.11+ (not in v1.10)?

File: pkg/apis/work/v1alpha1/well_known_constants.go [hunk 6]
```diff
@@ -16,19 +16,25 @@ limitations under the License.
...
 	ResourceBindingNamespaceLabel = "resourcebinding.karmada.io/namespace"
 
 	// ResourceBindingNameLabel is added to objects to specify associated ResourceBinding's name.
+	// Deprecated
 	ResourceBindingNameLabel = "resourcebinding.karmada.io/name"
 
 	// ClusterResourceBindingLabel is added to objects to specify associated ClusterResourceBinding.
+	// Deprecated
 	ClusterResourceBindingLabel = "clusterresourcebinding.karmada.io/name"
```
Comment (RainbowMango) [inline pkg/apis/work/v1alpha1/well_known_constants.go:31]: These unused labels, we can safely remove them.

File: pkg/apis/work/v1alpha1/well_known_constants.go (hunk 1 above)
Comment (RainbowMango) [reply pkg/apis/work/v1alpha1/well_known_constants.go:34]: I tend to don't touch these labels, let this PR focus on resourcebinding.karmada.io/key and clusterresourcebinding.karmada.io/key. But I'm ok with it.

**Comment by RainbowMango** [comment]
I updated the release notes a little bit, by the way.

File: pkg/apis/work/v1alpha1/well_known_constants.go (hunk 1 above)
Comment (whitewindmills) [reply pkg/apis/work/v1alpha1/well_known_constants.go:34]: fine, let's focus on rb/crb.

File: pkg/util/helper/work.go (hunk 2 above)
Comment (whitewindmills) [reply pkg/util/helper/work.go:120]: +1

File: pkg/controllers/binding/common.go (hunk 5 above)
Comment (whitewindmills) [reply pkg/controllers/binding/common.go:165]: yes

File: pkg/dependenciesdistributor/dependencies_distributor.go (hunk 3 above)
Comment (whitewindmills) [reply pkg/dependenciesdistributor/dependencies_distributor.go:528]: yes, it can only be created here.
```go
attachedBinding.Labels = util.DedupeAndMergeLabels(attachedBinding.Labels,
		map[string]string{workv1alpha2.ResourceBindingPermanentIDLabel: uuid.New().String()})
if err := d.Client.Create(context.TODO(), attachedBinding); err != nil {
```

File: pkg/dependenciesdistributor/dependencies_distributor.go (hunk 4 above)
Comment (whitewindmills) [reply pkg/dependenciesdistributor/dependencies_distributor.go:543]: it may occur when concurrently reconciling.

File: pkg/dependenciesdistributor/dependencies_distributor.go (hunk 3 above)
Comment (RainbowMango) [reply pkg/dependenciesdistributor/dependencies_distributor.go:528]: get it.

File: pkg/dependenciesdistributor/dependencies_distributor.go (hunk 4 above)
Comment (RainbowMango) [reply pkg/dependenciesdistributor/dependencies_distributor.go:543]: really? But we each reconciling only focus on 1 ResourceBinding. right?

Comment (RainbowMango) [reply pkg/dependenciesdistributor/dependencies_distributor.go:543]: Generally looks to me now, except this place.

Comment (whitewindmills) [reply pkg/dependenciesdistributor/dependencies_distributor.go:543]: Is it possible for a ResourceBinding to trigger multiple events at the same time?
https://github.com/karmada-io/karmada/blob/bad778bbb921a4c96f9d95d82a587a29d5684127/pkg/dependenciesdistributor/dependencies_distributor.go#L622-L625

I think so.

Comment (RainbowMango) [reply pkg/dependenciesdistributor/dependencies_distributor.go:543]: No, that won't happen. Each object can only be reconciled by one worker at the same time.
I can't find the documents right now. But this is the basis for a controller framework to guarantee.

Comment (whitewindmills) [reply pkg/dependenciesdistributor/dependencies_distributor.go:543]: Thanks for your remainder, I've found relevant instructions in the source code comments.
image
### Refer to
https://github.com/kubernetes-sigs/controller-runtime/blob/20f3f4bed925a9b7e41388c993fb217e7821910f/pkg/internal/controller/controller.go#L225-L227

Comment (RainbowMango) [reply pkg/dependenciesdistributor/dependencies_distributor.go]: Yes, exactly!

**Review by RainbowMango — APPROVED** [review]
/approve
Leave LGTM to @XiShanYongYe-Chang for another look.

**Review by XiShanYongYe-Chang — COMMENTED** [review]
Thanks a lot
/lgtm

**Comment by RainbowMango** [comment]
What's the next PR I can help with?

**Comment by RainbowMango** [comment]
Got it.
